runtime/src/seekable_buffering_reader.rs

```diff
@@ -0,0 +1,195 @@
+}
+
+impl Clone for SeekableBufferingReader {
+    fn clone(&self) -> Self {
+        Self {
+            instance: Arc::clone(&self.instance),
+            pos: 0,
+            last_buffer_index: 0,
+            next_index_within_last_buffer: 0,
+        }
+    }
+}
+
+impl Drop for SeekableBufferingReaderInner {
+    fn drop(&mut self) {
+        if let Some(handle) = self.bg_reader.lock().unwrap().take() {
+            self.stop.store(true, Ordering::Relaxed);
+            handle.join().unwrap();
+        }
+    }
+}
+
+impl SeekableBufferingReader {
+    pub fn new<T: 'static + Read + std::marker::Send>(reader: T) -> Self {
+        let inner = SeekableBufferingReaderInner {
+            new_data: RwLock::new(vec![]),
+            data: RwLock::new(vec![]),
+            len: AtomicUsize::new(0),
+            calls: AtomicUsize::new(0),
+            error: RwLock::new(Ok(0)),
+            bg_reader: Mutex::new(None),
+            stop: AtomicBool::new(false),
+        };
+        let result = Self {
+            instance: Arc::new(inner),
+            pos: 0,
+            last_buffer_index: 0,
+            next_index_within_last_buffer: 0,
+        };
+
+        let result_ = result.clone();
+
+        let handle = Builder::new()
+            .name("solana-compressed_file_reader".to_string())
+            .spawn(move || {
+                result_.read_entire_file_in_bg(reader);
+            });
+        *result.instance.bg_reader.lock().unwrap() = Some(handle.unwrap()); // TODO - unwrap here - do we expect to fail creating a thread? If we do, probably a fatal error anyway.
+        std::thread::sleep(std::time::Duration::from_millis(200)); // hack: give time for file to be read a little bit
+        result
+    }
+    fn read_entire_file_in_bg<T: 'static + Read + std::marker::Send>(&self, mut reader: T) {
+        let mut time = Measure::start("");
+        const CHUNK_SIZE: usize = 65536 * 2;
+        let mut data = [0u8; CHUNK_SIZE];
+        loop {
+            if self.instance.stop.load(Ordering::Relaxed) {
+                info!("stopped before file reading was finished");
+                break;
+            }
+            let result = reader.read(&mut data);
+            match result {
+                Ok(size) => {
+                    self.instance
+                        .new_data
+                        .write()
+                        .unwrap()
+                        .push(data[0..size].to_vec());
+                    self.instance.len.fetch_add(size, Ordering::Relaxed);
+                    if size == 0 {
+                        break;
+                    }
+                }
+                Err(err) => {
+                    error!("error reading file");
+                    *self.instance.error.write().unwrap() = Err(err);
+                    break;
+                }
+            }
+        }
+        time.stop();
+        error!(
+            "reading entire decompressed file took: {} us, bytes: {}",
```

Comment:
is it an error?

Reply:
no, still prototype code. just how I log when testing so things always show up.